starlette/testclient.py

```diff
@@ -152,11 +177,40 @@ async def _asgi_receive(self) -> Message:
     async def _asgi_send(self, message: Message) -> None:
         self._send_queue.put(message)
 
-    def _raise_on_close(self, message: Message) -> None:
+    def _handle_response(self, message: Message) -> None:
+        status_code: int = message["status"]
+        headers: typing.List[typing.Tuple[bytes, bytes]] = message["headers"]
+        body = []
+        while True:
+            message = self.receive()
+            assert message["type"] == "websocket.http.response.body"
+            body.append(message["body"])
+            if not message.get("more_body", False):
+                break
+        raise WebSocketReject(
+            response_status=status_code,
+            response_headers=headers,
+            response_body=b"".join(body),
+        )
+
+    def _raise_on_close(
+        self, message: Message, reject: typing.Optional[bool] = None
+    ) -> None:
         if message["type"] == "websocket.close":
-            raise WebSocketDisconnect(
-                message.get("code", 1000), message.get("reason", "")
-            )
+            if not reject:
+                raise WebSocketDisconnect(
+                    message.get("code", 1000), message.get("reason", "")
+                )
+            else:
+                # A webserver which gets a "close" before "accept"
+                # will return a 403 response.  It may or may not do anything
+                # with the close code and reason.  We store it in this exception
+                # for test introspection.
```

Comment:
If the `code` and `reason` are not being used, why the `WebSocketReject` would store it?

I think this is more confusing.

Reply:
So that tests, using the TestClient, can verify that the _application_ sent the code and reason.
This is a TestClient, designed to test if the application is behaving in a certain way.  If I write my websocket app, and return a close and reason, in my unit tests I want to make sure that that close and reason are correctly generated.

Reply:
WebSocketReject() (now WebSocketDenialResponse) was previously sent both for close() type rejection and also for http.response type rejection.  The idea was that this class would capture as much as possible about the rejection event, for the TestClient to be able to see how the application performed its connection rejection.

Latest change no longer sends a WebSocketDenialResponse() when a close() type rejection is performed.